lib/ansible/modules/source_control/gitlab_service.py

```diff
@@ -0,0 +1,377 @@
+'''
+
+
+def init_definitions():
+    sd = json.loads(RAW_SERVICES_DEFINITIONS)
+    for service_slug, params in sd.items():
+        for name, definition in params.items():
+            if 'type' in definition:
+                definition['type'] = bool if definition['type'] == 'bool' else int if definition['type'] == 'int' else definition['type']
+            if name in GitLabServices.credentialParams:
+                definition['no_log'] = True
+    return sd
+
+
+class GitLabServices(object):
+    HOOK_EVENTS = ['push', 'issues', 'confidential_issues', 'merge_requests', 'tag_push', 'note', 'confidential_note', 'job', 'pipeline', 'wiki_page']
+    credentialParams = ['password', 'token', 'api_key']
+
+    def __init__(self, module, git):
+        self._module = module
+
+    def update(self, project, remote, active, params, events):
+        diff = not self.equals(remote.attributes, active, params, events)
+        if diff:
+            # remote.active = active
+            for k, v in self.expandEvents(events).items():
+                setattr(remote, k, v)
+            # when saving, params does directly in the request
+            # (remember they get retrieved under the "property" key of the remote dict())
+            for k, v in params.items():
+                setattr(remote, k, v)
+            if self._module.check_mode:
+                self._module.exit_json(changed=True)
+            remote.save()
+            return remote
+        else:
+            return False
+
+    def asApiObject(self, active, params, events):
```

Comment:
Why not using directly `project.services.create` method ?

Reply:
As far as I remember I had some trouble using `create` without having the tweak the `dict()/data-structure` in use.
And since we always do `remote_service = project.services.get(service)`, we always have a `remote service` object suitable for `project.services.save`.